File: data_structures/c++/MergeSort.cpp

#include <iostream>
#include <string>
#include <algorithm>

using namespace std;

class  MergeSort{
    public:
    
    static void sort(int* arr, int len) {
        int* copy = new int[len];
        split(arr, 0, len, copy);
    }

    static void sort_iterative(int* arr, int len) {
        int* copy = new int[len];
        split_iterative(arr, 0, len, copy);
    }
    
    private:

    static void split(int* arr, int begin, int end, int* copy) {
        if((end-begin) < 2) {
            return;
        }
        
        int middle = (begin + end) / 2;
        split(arr, begin, middle, copy);
        split(arr, middle, end, copy);
        merge(arr, begin, middle, end, copy);
        copy_back(copy, begin, end, arr);    
    }

    static void split_iterative(int* arr, int begin, int end, int* copy) {
        for(int width = 1; width < end; width = width * 2) {
            for(int i = 0; i < end; i = i + (width * 2)) {
                merge(arr, i, min(i + width, end), min(end, i + (width * 2)), copy);
            }
            copy_back(copy, begin, end, arr);
        }
    }
    
    static void merge(int * arr, int begin, int middle, int end, int * copy) {
        int ptr_1 = begin;
        int ptr_2 = middle;
        
        for(int i = begin; i < end; i++) {
            if(ptr_1<middle && (ptr_2 >= end || arr[ptr_1] <= arr[ptr_2])) {
                copy[i] = arr[ptr_1];
                ptr_1 += 1;
            } else {
                copy[i] = arr[ptr_2];
                ptr_2 += 1;
            }
        }
    }
    
    static void copy_back(int * copy, int begin, int end, int * arr) {
        for(int i = begin; i < end; i++) {
            arr[i] = copy[i];
        }
    }
};

int main()
{
    int n, r;
    string x;
    cout << "Enter N: ";
    cin >> n;
    cout << "Recursive Algorithm?: ";
    cin >> r;
    int * arr = new int[n];
    for(int i = 0; i < n; i++) {
        cin >> x;
        arr[i] = stoi(x);
    }
    cout << endl << endl;
    if(r) {
        cout << "Recursive" << endl;
        MergeSort::sort(arr, n);
    }
    else
        MergeSort::sort_iterative(arr, n);

    for(int i = 0; i < n; i++) {
        cout << arr[i] << "    ";
    }
    cout << endl;
}

File: algorithms/sorts/MergeSort.java
public class MergeSort {

	/**
	 * The main sorting function
	 * Time - O(nlog(n))
	 * Space - O(n)
	 */
	public static void sort(int [] A, int [] B, int n, boolean recursive) {
		if(recursive)
			split(A, 0, n , B);
		else
			splitIterative(A,n,B);
	}

	/**
	 * Recursive function that splits the array - O(log(n))
	 */
	public static void split(int [] A, int begin, int end, int [] B) {
		if(end-begin < 2) {
			return;
		}

		int middle = (begin + end)/2;
		split(A, begin, middle, B);
		split(A, middle, end, B);
		merge(A, begin, middle, end, B);
		copy(B, begin, end, A);
	}

	/**
	 * Iterative function to split the array - O(log(n))
	 */
	public static void splitIterative(int [] A, int n, int [] B) {
		for(int width = 1; width < n; width = width * 2) {
			for(int i = 0; i < n; i = i + width * 2) {
				merge(A, i, Math.min(i + width, n),Math.min(i + 2 * width, n), B);
			}
			copy(B, 0, n, A);
		}
	}

	/**
	 * The merging function - O(n)
	 */
	public static void merge(int [] A, int begin, int middle, int end, int [] B) {
		//Pointers pointing at current elements in the two arrays
		int p1 = begin, p2 = middle;
		for(int i = begin; i < end; i++) {
			//Check if pointer 1 is less than middle
			//and 2 hasn't reached the end
			//and value at pointer 1 is greater than element at pointer 2
			if(p1 < middle && (p2 >= end || A[p1] <= A[p2])) {
				B[i] = A[p1];
				p1 = p1 + 1;
			} else {
				B[i] = A[p2];
				p2 = p2 + 1;
			}
		}
	}

	/**
	 * Copy array values back
	 */
	public static void copy(int [] B, int begin, int end, int [] A) {
		for(int i = begin; i < end; i++) {
			A[i] = B[i];
		}
	}

	public static void main(String [] args) {
		int [] arr = new int [args.length];
		for(int i = 0; i < args.length - 1; i++) {
			arr[i] = Integer.parseInt(args[i]);
		}
		int [] arr2 = new int[arr.length-1];
		if(Integer.parseInt(args[args.length-1]) != 0) {
			sort(arr, arr2, arr.length-1, true);
		} else {
			sort(arr, arr2, arr.length-1, false);
		}

		System.out.println(java.util.Arrays.toString(arr));
	}
}